lib/buffer.js

```diff
@@ -246,18 +247,30 @@ function fromArrayLike(obj) {
   const length = obj.length;
   const b = allocate(length);
   for (var i = 0; i < length; i++)
-    b[i] = obj[i] & 255;
+    b[i] = obj[i];
   return b;
 }
 
 function fromArrayBuffer(obj, byteOffset, length) {
+  if (!isArrayBuffer(obj))
+    throw new TypeError('argument is not an ArrayBuffer');
+
   byteOffset >>>= 0;
 
-  if (typeof length === 'undefined')
-    return binding.createFromArrayBuffer(obj, byteOffset);
+  const maxLength = obj.byteLength - byteOffset;
+
+  if (maxLength <= 0)
+    throw new RangeError("'offset' is out of bounds");
+
+  if (length === undefined) {
+    length = maxLength;
+  } else {
+    length >>>= 0;
+    if (length > maxLength)
+      throw new RangeError("'length' is out of bounds");
+  }
 
-  length >>>= 0;
-  return binding.createFromArrayBuffer(obj, byteOffset, length);
+  return new FastBuffer(obj, byteOffset, length);
```

Comment:
Now that you're directly extending Uint8Array, the checks above would only seem necessary if we're trying to preserve the error messages. Otherwise the Uint8Array constructor would catch it.